### ISSUE TITLE
Can't deploy fully configured MetalLB via ArgoCD via helm

### ISSUE THREAD

**Issue by dusatvoj** [metallb/metallb#2241]
### MetalLB Version

0.13.12

### Deployment method

Charts

### Main CNI

Cillium

### Kubernetes Version

1.27.9

### Cluster Distribution

rancher

### Describe the bug

I want to use ArgoCD as configurator of whole cluster so I have per-cluster helm chart with argocd apps. All apps so far can be completely configured via helm chart values.

MetalLB can't be fully configured. After creating the whole setup I have tu run raw kubernetes manifest with `IPAddressPool` and `L2Advertisement`.

I found out there's bug describing the same but it's closed (https://github.com/metallb/metallb/issues/1691) and also closed but unmerged PR (https://github.com/metallb/metallb/pull/1484).

### To Reproduce

1. Deploy emtpy cluster
2. Deploy argocd with main helmchart
3. Everything works but MetalLB (as critical component) doesn't have defined `IPAddressPool` and `L2Advertisement`
4. _You have to manually deploy these manifests_

### Expected Behavior

Can define `IPAddressPool` and `L2Advertisement` (and probably `BGPAdvertisement` ) via helm values as optional object

### Additional Context

_I just submitted everything below just for allowing me to submit this issue. The closed issue has ignored comments for months, so I don't want to be just +1 ignored comment. I think this is important thing to be completed._

### I've read and agree with the following

- [X] I've checked all open and closed issues and my request is not there.
- [X] I've checked all open and closed pull requests and my request is not there.

### I've read and agree with the following

- [X] I've checked all open and closed issues and my issue is not there.
- [X] This bug is reproducible when deploying MetalLB from the main branch
- [X] I have read the [troubleshooting guide](https://metallb.universe.tf/troubleshooting/) and I am still not able to make it work
- [X] I checked the logs and MetalLB is not discarding the configuration as not valid
- [X] I enabled the debug logs, collected the information required from the cluster using the [collect script](https://raw.githubusercontent.com/metallb/metallb/main/troubleshooting/collect.sh) and will attach them to the issue
- [X] I will provide the definition of my service and the related endpoint slices and attach them to this issue

**Comment by fedepaol** [comment]
Thanks for raising this again!

We just need to understand what is the canonical way for helm to handle this / how other projects are handling it.

I am not super helm-savvy. @dusatvoj do you have references of other charts that allow embedding the CRs directly with helm? Otherwise we'll try to look for answers, but I don't feel allowing a blob in our chart is the right way (I might be wrong).

**Comment by dusatvoj** [comment]
Not sure, maybe there could be a list with ipAddressPools, l2Advertisement and bgpAdvertisement and in foreach there could be generated CRs.

**Comment by fedepaol** [comment]
I'll try to do an assessment on other projects and will ask on the helm channel on slack. I just want to avoid doing anything too esotic.

**Comment by dusatvoj** [comment]
@tamcore , good point - thank you for suggestion but still I think this is just workaround, bcs MetalLB should have option to be configured via it's own helm chart :thinking:

**Comment by disco-stu** [comment]
> Helm has a special `crds` folder. That way helm will install the CRDs before other any other resources: https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you

Thats what i meant by "treat CRDs as CRDs in the helm way".

**Comment by dusatvoj** [comment]
@disco-stu , many thank's for taking care of this issue and also explanation.

So what are next steps? Close this issue and keep it as "how to"? :thinking:

**Comment by disco-stu** [comment]
@dusatvoj in my opinion we should try to come up with a dedicated metallb-config helm chart. Even if it will not solve the initial issue in the first place, it prevents uses from having to write there own helm charts or additional configs.

Later on, if one decides to use the usual way in helm to deploy CRDs, we could reuse that config helm chart and reuse it.

I would be happy to support and would try to come up with such a metallb-config helm chart.

**Comment by fedepaol** [comment]
> Helm has a special `crds` folder. That way helm will install the CRDs before other any other resources: https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you

The problem with this is "These CRDs are not templated," while because of the conversion webhook we must inject the namespace [here](https://github.com/metallb/metallb/blob/main/charts/metallb/charts/crds/templates/crds.yaml#L338)

**Comment by fedepaol** [comment]
> @dusatvoj in my opinion we should try to come up with a dedicated metallb-config helm chart. Even if it will not solve the initial issue in the first place, it prevents uses from having to write there own helm charts or additional configs.
> 
> Later on, if one decides to use the usual way in helm to deploy CRDs, we could reuse that config helm chart and reuse it.
> 
> I would be happy to support and would try to come up with such a metallb-config helm chart.

@disco-stu if I understand your suggestion is to separate the config part from the CRD part and have an umbrella chart that may contain both?

**Comment by disco-stu** [comment]
@fedepaol not exactly. I've already created a draft for a new "standalone" helm chart. https://github.com/credativ/metallb/tree/metallb-config

The metallb-config chart could be used to configure metallb after it's first installation.

**Comment by fedepaol** [comment]
> > > Helm has a special `crds` folder. That way helm will install the CRDs before other any other resources: https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you
> > 
> > 
> > The problem with this is "These CRDs are not templated," while because of the conversion webhook we must inject the namespace [here](https://github.com/metallb/metallb/blob/main/charts/metallb/charts/crds/templates/crds.yaml#L338)
> 
> Then i would suggest to have a look at how other projects do it. ingress-nginx for example, deploys a Job that patches the their webhook: https://github.com/kubernetes/ingress-nginx/blob/main/charts/ingress-nginx/templates/admission-webhooks/job-patch/job-patchWebhook.yaml.
> 
> Other projects, like Velero, have a similar approach: https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/templates/upgrade-crds/upgrade-crds.yaml

Interesting! But we'd still have the issue of crds not upgrading automatically right?

**Comment by disco-stu** [comment]
> > > > Helm has a special `crds` folder. That way helm will install the CRDs before other any other resources: https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you
> > > 
> > > 
> > > The problem with this is "These CRDs are not templated," while because of the conversion webhook we must inject the namespace [here](https://github.com/metallb/metallb/blob/main/charts/metallb/charts/crds/templates/crds.yaml#L338)
> > 
> > 
> > Then i would suggest to have a look at how other projects do it. ingress-nginx for example, deploys a Job that patches the their webhook: https://github.com/kubernetes/ingress-nginx/blob/main/charts/ingress-nginx/templates/admission-webhooks/job-patch/job-patchWebhook.yaml.
> > Other projects, like Velero, have a similar approach: https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/templates/upgrade-crds/upgrade-crds.yaml
> 
> Interesting! But we'd still have the issue of crds not upgrading automatically right?

The solution that velero implemented overcomes this issue by running a pre-install,pre-upgrade,pre-rollback hook that manually applies the CRDs. It uses the velero binary to get the right version of those CRDs and applies those with the help of kubectl.

This would also solve this issue, but in my opinion it feels a little bit hacky.

**Comment by disco-stu** [comment]
IMHO we have two reasonable options:

1. Use a two step approach: Use the metallb-chart as it is right now and use a second chart to configure metalllb (thats what I'm working on right now).

2. Build a POC based on the velero and nginx-ingress approach. This might include the following changes: Move the CRDs into the Helm crds folder, which would include removing all templated variables. Within a pre-xxx hook: Patch the webhook namespace to point to the right webhook receiver. This pre-xxx hook job could also update the CRDs to overcome the helm upgrade CRD issue.

However, even if solution 2 would solve the various issues: template in crds, update crds on helm upgrade and maybe rollback, we would still have the problem on a "NotReady" webhook pod at installation time, if we want to deploy CRs in the same run.

Any thoughts? Did I missed something?

**Comment by dylannorthrup** [comment]
As someone who is not using Argo but **would** like to manage my metallb installs via helm, having a supported path forward just using the provided helm charts for this would be good. I am hopeful a decision can be made between `metallb-config` branch from @disco-stu and the velero/nginxingress model; and the path forward can be made clear for non-Argo `helm` users.

**Comment by rastaman** [comment]
In my case it looks like it works with the`"helm.sh/hook": post-install ` to create the resources:

```yaml
---
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: default
  annotations:
    "helm.sh/hook": post-install <---- install this after other stuff
spec:
  addresses:
  - {{ .Values.cidr }}
  avoidBuggyIPs: true
status: {}
---
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: l2advertisement1
  annotations:
    "helm.sh/hook": post-install
spec:
  ipAddressPools:
  - default
status: {}
```

However i don't install it every day so perhaps my test is not enough :-)
